drivers/i3c/i3c_npcx.c

```diff
@@ -0,0 +1,2053 @@
+	static struct i3c_device_desc npcx_i3c_device_array_##id[] = I3C_DEVICE_ARRAY_DT_INST(id); \
+	static struct i3c_i2c_device_desc npcx_i3c_i2c_device_array_##id[] =                       \
+		I3C_I2C_DEVICE_ARRAY_DT_INST(id);                                                  \
+	static const struct npcx_i3c_config npcx_i3c_config_##id = {                               \
+		.base = (struct i3c_reg *)DT_INST_REG_ADDR(id),                                    \
+		.clock_dev = DEVICE_DT_GET(NPCX_CLK_CTRL_NODE),                                    \
+		.reset = RESET_DT_SPEC_INST_GET(id),                                               \
+		.clock_subsys = NPCX_DT_CLK_CFG_ITEM_BY_NAME(id, mclkd),                           \
+		.ref_clk_subsys = NPCX_DT_CLK_CFG_ITEM_BY_NAME(id, apb4),                          \
+		.irq_config_func = npcx_i3c_config_func_##id,                                      \
+		.common.dev_list.i3c = npcx_i3c_device_array_##id,                                 \
+		.common.dev_list.num_i3c = ARRAY_SIZE(npcx_i3c_device_array_##id),                 \
+		.common.dev_list.i2c = npcx_i3c_i2c_device_array_##id,                             \
+		.common.dev_list.num_i2c = ARRAY_SIZE(npcx_i3c_i2c_device_array_##id),             \
+		.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(id),                                      \
+		.clocks.i3c_pp_scl_hz = DT_INST_PROP_OR(id, i3c_scl_hz, 0),                        \
+		.clocks.i3c_od_scl_hz = DT_INST_PROP_OR(id, i3c_od_scl_hz, 0),                     \
+		IF_ENABLED(CONFIG_I3C_NPCX_DMA, (                                                  \
+			.mdma_clk_subsys = NPCX_DT_CLK_CFG_ITEM_BY_IDX(id, 2),                     \
+		))                                                                                 \
+		IF_ENABLED(CONFIG_I3C_NPCX_DMA, (                                                  \
+			.mdma_base = (struct mdma_reg *)DT_INST_REG_ADDR_BY_IDX(id, 1),            \
+		))                                                                                 \
+	};                                                                                         \
+	static struct npcx_i3c_data npcx_i3c_data_##id;                                            \
+	DEVICE_DT_INST_DEFINE(id, npcx_i3c_init, NULL, &npcx_i3c_data_##id, &npcx_i3c_config_##id, \
+			      POST_KERNEL, CONFIG_I3C_CONTROLLER_INIT_PRIORITY,                    \
+			      &npcx_i3c_driver_api);                                               \
+	static void npcx_i3c_config_func_##id(const struct device *dev)                            \
+	{                                                                                          \
+		IRQ_CONNECT(DT_INST_IRQN(id), DT_INST_IRQ(id, priority), npcx_i3c_isr,             \
+			    DEVICE_DT_INST_GET(id), 0);                                            \
+		irq_enable(DT_INST_IRQN(id));                                                      \
+	};
```

Comment:
you could move this to the top of the macro, then you won't have to forward declare it

Reply:
Done